### PR TITLE
[SPARK-44971][PYTHON] StreamingQueryProgress event fromJson bug fix

### PR DESCRIPTION
**Pull request by WweiL**
### What changes were proposed in this pull request?

The `fromJson` method for `StreamingQueryProgress` excepts the field `batchDuration` is in the dict.
That method is used internally for converting a json representation of `StreamingQueryProgress` into python object, commonly created in the Scala side `json` method of the same object.

But the `batchDuration` field is not there before https://github.com/apache/spark/pull/42077, which is only merged to 4.0. Therefore we add a catch there to prevent this method from failing.

### Why are the changes needed?

Necessary bug fix

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Added unit test

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
Thank you, @WweiL . SPARK-44971 seems to be filed as a new feature (maybe by default value). You can switch to `Bug` issue type.

Screenshot 2023-08-25 at 11 51 54 AM

File: python/pyspark/sql/streaming/listener.py [hunk 1]
```diff
@@ -477,7 +477,7 @@ def fromJson(cls, j: Dict[str, Any]) -> "StreamingQueryProgress":
             name=j["name"],
             timestamp=j["timestamp"],
             batchId=j["batchId"],
-            batchDuration=j["batchDuration"],
+            batchDuration=j["batchDuration"] if "batchDuration" in j else None,
```
Comment (dongjoon-hyun) [inline python/pyspark/sql/streaming/listener.py:480]: I'm wondering when Spark receive a JSON file without `batchDuration`? We are inside the same Spark versions, aren't we? Is this listener able to listen old Spark?

Comment (WweiL) [reply python/pyspark/sql/streaming/listener.py:480]: So this is currently how this method used: in spark connect, the way the listener works, is that 
1. The user's listener code is serialized and sent to spark server
2. The server starts a scala listener, [in which starts a python process](https://github.com/apache/spark/blob/master/connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/StreamingQueryListenerHelper.scala) (essentially as another connect client), that runs the user's code
3. Each time a new event comes in, the event on [java side is serialized to json](https://github.com/apache/spark/blob/master/connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/StreamingQueryListenerHelper.scala#L52) and passed to server python process, which calls this `fromJson` [method to convert it back to the actual `StreamingQueryProgress` object](https://github.com/apache/spark/blob/master/python/pyspark/sql/connect/streaming/worker/listener_worker.py#L76-L77)

But before https://github.com/apache/spark/pull/42077 in 3.5, that field is not added in the jvm `json` method of `StreamingQueryProgress`. Here it excepts that to always be presented, hence we get an error

**Review by HyukjinKwon — APPROVED** [review]

File: python/pyspark/sql/streaming/listener.py (hunk 1 above)
Comment (HyukjinKwon) [inline python/pyspark/sql/streaming/listener.py:480]:
```suggestion
            batchDuration=j.get("batchDuration", None),
```

**Comment by WweiL** [comment]
Hi @HyukjinKwon can we merge this : )

**Comment by HyukjinKwon** [comment]
Merged to master and branch-3.5.

**Comment by WweiL** [comment]
@HyukjinKwon Can we just merge this to 3.5

**Comment by WweiL** [comment]
ah nvm I set the merge target to 3.5, should be fine

**Comment by HyukjinKwon** [comment]
Yup, merged to 3.5.

**Pull request closed without merging**